sklearn/metrics/_regression.py

```diff
@@ -137,6 +140,63 @@ def _check_reg_targets(y_true, y_pred, multioutput, dtype="numeric", xp=None):
     return y_type, y_true, y_pred, multioutput
 
 
+def _check_reg_targets_with_floating_dtype(
+    y_true, y_pred, sample_weight, multioutput, xp=None
+):
+    """Ensure that y_true and y_pred correspond to the same regression task.
+
+    This helper automatically selects a floating-point data type with
+    `_find_matching_floating_dtype` when computing with inputs that
+    follow the array API specification and ensure that all non-None inputs
+    are converted to that common dtype if needed before returning them.
+
+    Parameters
+    ----------
+    y_true : array-like
+
+    y_pred : array-like
+
+    sample_weight : array-like of shape (n_samples,)
+
+    multioutput : array-like or string in ['raw_values', 'uniform_average', \
+        'variance_weighted'] or None
+        None is accepted due to backward compatibility of r2_score().
+
+    xp : module, default=None
+        Precomputed array namespace module. When passed, typically from a caller
+        that has already performed inspection of its own inputs, skips array
+        namespace inspection.
+
+    Returns
+    -------
+    type_true : one of {'continuous', 'continuous-multioutput'}
+        The type of the true target data, as output by
+        'utils.multiclass.type_of_target'.
+
+    y_true : array-like of shape (n_samples, n_outputs)
+        Ground truth (correct) target values.
+
+    y_pred : array-like of shape (n_samples, n_outputs)
+        Estimated target values.
+
+    sample_weight : array-like of shape (n_samples,), default=None
+        Sample weights.
+
+    multioutput : array-like of shape (n_outputs) or string in ['raw_values', \
+        'uniform_average', 'variance_weighted'] or None
+        Custom output weights if ``multioutput`` is array-like or
+        just the corresponding argument if ``multioutput`` is a
+        correct keyword.
+    """
+    dtype_name = _find_matching_floating_dtype(y_true, y_pred, sample_weight, xp=xp)
+
+    y_type, y_true, y_pred, multioutput = _check_reg_targets(
+        y_true, y_pred, multioutput, dtype=dtype_name, xp=xp
+    )
```

Comment:
I think we need to add a line that converts `sample_weight` to the shared dtype no? At least that was kind of my expectation (also later in this PR we remove a line that will adjust the dtype of `sample_weight`), so if we don't convert it maybe we should add a comment to explain why. That way people from the future will have an easier time.

Reply:
I must've switched my brain off when I wrote that! 😅  Thanks for catching it!